### PR TITLE
tests for user api 

### PR DESCRIPTION
**Pull request by mpieron**

### CODE REVIEW

File: Vikta-API-testing-sample/src/main/java/com/test/api/dto/UserDTO.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+public class UserDTO {
+    private Integer id;
+    private String loginName;
+    private String password;
+    private String email;
+    private String firstName;
+    private String middleName;
+    private String surname;
+    private String pathToAvatarImage;
+    private int[] addressIds;
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/main/java/com/test/api/dto/UserDTO.java:15]: I think it would be easier to test if arrays were replaced by Lists

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                .as("There should be no users with same login")
+                .isGreaterThan(0);
+
+        assertThat(resultList)
+                .as("Found user with wrong surname")
+                .allMatch(user -> user.getSurname().equals(userSurname));
+    }
+
+    private List<UserDTO> getListOfUsers(){
+        String pathToList = getTestEnvironment().getUserPath() + "/list";
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:183]: add usersPath property in test.properties - just as categoriesPath and refactor all use cases to get rid of using '+"\list"\' in tests

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+                .get(loginPath)
+                .then()
+                .log().all()
+                .spec(defaultResponseSpec())
+                .extract()
+                .body().jsonPath().getList(".", UserDTO.class);
+
+        assertThat(user.size())
+                .as("There should be no users with same login")
+                .isLessThan(2);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:86]: I think it would be better to assert equalTo(1)

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+    @Test
+    public void canGetUserByLoginName(){
+        final String loginPath = getTestEnvironment().getUserPath() + "/login";
+        final String loginQueryParam = "login";
+
+        List<UserDTO> users = getListOfUsers();
+        String userLogin = users.get(new Random().nextInt(users.size())).getLoginName();
+
+        List<UserDTO> user = given().spec(defaultRequestSpec())
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:73]: change the name of the variable to something more fit to List

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                .as(String.format("Should be user with id %s, but was with %s", id, user.getId()))
+                .isEqualTo(id);
+    }
+
+    @Test
+    public void canListAllUsers(){
+        List<UserDTO> users = getListOfUsers();
+
+        assertThat(users).as("User list shouldn't be empty").isNotEmpty();
+        assertThat(users).as("User don't have loginName field").allMatch(user -> !user.getLoginName().isEmpty());
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:57]: please use allSatisfy() assertion for that - will discuss on call

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        assertThat(user.size())
+                .as("There should be no users with same login")
+                .isLessThan(2);
+
+        assertThat(user.get(0).getLoginName())
+                .as(String.format("Should be user with login %s, but was with %s", userLogin, user.get(0).getLoginName()))
+                .isEqualTo(userLogin);
+    }
+
+    // when user is not found (does not exist) returns empty body, I think that better will be returned code 404 - not found
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:93]: add that to list of things to report

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    @Test
+    public void canSearchUserByTerm(){
+        final String searchPath = getTestEnvironment().getUserPath() + "/search";
+        final String termQueryParam = "term";
+        Random random = new Random();
+        List<UserDTO> users = getListOfUsers();
+
+        assertThat(users.size()).isGreaterThan(0);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:121]: assert that list is not empty

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+        int size = random.nextInt(users.size()-1) + 2;
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:123]: you don't have to substract 1. nextInt method picks a value between 0 (inclusive) and the given bound (exclusive)

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                .queryParam(termQueryParam, term)
+                .when()
+                .log().all()
+                .get(searchPath)
+                .then()
+                .log().all()
+                .spec(defaultResponseSpec())
+                .extract().body().jsonPath().getList(".", UserDTO.class);
+
+        assertThat(users)
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:141]: you already asserted that at the beginning of the test. keep one assertion

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]
add all used paths to test.properties and get them from there. check the docs to see if you can test some boundary values with parameterized tests

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+    public void canListAllUsers(){
+        List<UserDTO> usersList = getListOfUsers();
+
+        assertThat(usersList).as("User list shouldn't be empty").isNotEmpty();
+
+        assertThat(usersList).as("Fields shouldn't be null").allSatisfy(user -> assertThat(user)
+                        .hasNoNullFieldsOrProperties());
+
+        assertThat(usersList).as("Fields shouldn't be empty").allSatisfy(user -> assertThat(user)
+                .usingRecursiveComparison().isNotEqualTo(""));
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:91]: here you are trying to check equality between UserDTO object (user) and String (""). If you want to assert that received users don't have empty fields, you can check against UserDTO object that have all String fields set to "".

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 11]
```diff
@@ -0,0 +1,446 @@
...
+                .extract()
+                .body().jsonPath().getList(".", UserDTO.class);
+    }
+
+    private static Stream<Arguments> goodParametersForPost() {
+        String words64 = RandomStringUtils.randomAlphabetic(64);
+        String words63 = RandomStringUtils.randomAlphabetic(63);
+        String words3 = RandomStringUtils.randomAlphabetic(59);
+        // this is max that I can apply, 319 length string, more won't pass
+        String maxEmail = String.format("%s@%s.%s.%s.%s.com", words64, words63, words63, words63, words3);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:401]: do we need that many string parts? maybe create one longer string for username and then one shorter for domain name?

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 12]
```diff
@@ -0,0 +1,446 @@
...
+                Arguments.of("jan","password123", "email@123.a", "Jan", RandomStringUtils.randomAlphabetic(288), "Kowalski", "avatar", new int[]{}, new int[]{}, "Case with min value of middleName length"),
+                Arguments.of("jan","password123", "email@123.a", "Jan", "Janusz", RandomStringUtils.randomAlphabetic(1), "avatar", new int[]{}, new int[]{}, "Case with min value of surname length"),
+                Arguments.of("jan","password123", "email@123.a", "Jan", "Janusz", RandomStringUtils.randomAlphabetic(288), "avatar", new int[]{}, new int[]{}, "Case with min value of surname length")
+        );
+    }
+
+    private static Stream<Arguments> wrongParametersForPost() {
+        String words64 = RandomStringUtils.randomAlphabetic(64);
+        String words63 = RandomStringUtils.randomAlphabetic(63);
+        String maxEmail = String.format("%s@%s.%s.%s.%s.%s.com", words64, words63, words63, words63, words63, words63);
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:422]: same as above

File: Vikta-API-testing-sample/src/test/resources/test.properties [hunk 13]
```diff
@@ -3,3 +3,8 @@ port=5054
 basePath=/api/v1/
 categoryPath=category
 categoriesPath=${categoryPath}/list
+userPath=user
+usersListPath=${userPath}/list
+userLoginPath=${userPath}/login
+userSearchPath=${userPath}/search
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/resources/test.properties:9]: please do the same for category and rework the category test to use it

**Review by kdyla-griddynamics — CHANGES_REQUESTED** [review]

File: Vikta-API-testing-sample/src/test/resources/test.properties (hunk 13 above)
Comment (mpieron) [reply Vikta-API-testing-sample/src/test/resources/test.properties:9]: It's done but at main branch

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java (hunk 10 above)
Comment (mpieron) [reply Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:91]: so now is okey?
Screenshot 2022-02-01 at 3 42 48 PM

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java (hunk 11 above)
Comment (mpieron) [reply Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:401]: I checked that between dots can be max 63 chars(64 including dot), that's why here is so many string parts, but I can just check smaller string

File: Vikta-API-testing-sample/src/test/resources/test.properties (hunk 13 above)
Comment (kdyla-griddynamics) [reply Vikta-API-testing-sample/src/test/resources/test.properties:9]: ok

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java (hunk 11 above)
Comment (kdyla-griddynamics) [reply Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:401]: I think that checking against the most common email address pattern will be enough

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java (hunk 10 above)
Comment (kdyla-griddynamics) [reply Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:91]: looks good

File: Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+                .get(getTestEnvironment().getUsersListPath())
+                .then()
+                .spec(defaultResponseSpec())
+                .extract()
+                .body().jsonPath().getList(".", UserDTO.class);
+    }
+
+    private static Stream<Arguments> goodParametersForPost() {
+        String user = RandomStringUtils.randomAlphabetic(20);
+        String[] domains = new String[]{"@gmail.com", "opoczta.pl", "@op.pl", "poczta.onet.pl"};
```
Comment (kdyla-griddynamics) [inline Vikta-API-testing-sample/src/test/java/com/test/api/user/UserTest.java:406]: add @ to all domains

**Review by kdyla-griddynamics — APPROVED** [review]
